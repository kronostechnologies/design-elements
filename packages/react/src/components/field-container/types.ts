import { TooltipProps } from '../tooltip/tooltip';

export interface FieldSlotIds {
    label?: string;
    hint?: string;
    invalid?: string;
}

export interface FieldControlProps {
    id?: string;
    ariaLabel?: string;
    ariaLabelledby?: string;
    ariaDescribedby?: string
    disabled?: boolean;
    required?: boolean;
    valid?: boolean;
    slotIds?: FieldSlotIds;
}

export interface FieldContainerProps extends FieldControlProps {
    fieldId?: string;
    className?: string;
    noMargin?: boolean;
    label?: string;
    hint?: string;
    validationErrorMessage: string;
    noInvalidFieldIcon?: boolean;
    tooltip?: TooltipProps;
}
