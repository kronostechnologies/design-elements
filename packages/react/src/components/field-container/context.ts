import { createContext, useContext } from 'react';
import { FieldControlProps } from './types';

export const FieldControlContext = createContext<FieldControlProps>({});

export function useFieldControlContext(newContext: FieldControlProps): FieldControlProps {
    const prevContext = useContext(FieldControlContext);

    return {
        ...prevContext,
        ...newContext,
    };
}
