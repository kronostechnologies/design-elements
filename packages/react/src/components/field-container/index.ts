export { FieldContainer } from './field-container';
export type { FieldContainerProps } from './field-container';
