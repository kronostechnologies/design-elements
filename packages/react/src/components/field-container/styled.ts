import styled from 'styled-components';

interface StyledDivProps {
    $hasLabel: boolean;
    $hasHint: boolean;
    $valid: boolean;
    $noMargin?: boolean;
}

export const StyledDiv = styled.div<StyledDivProps>`
    margin: ${({ $noMargin }) => ($noMargin ? '0' : '0 0 var(--spacing-3x)')};

    input,
    select,
    textarea {
        border-color: ${({ theme, $valid }) => ($valid ? theme.component['field-input-border-color'] : theme.component['field-input-error-border-color'])};
    }

    > :nth-child(${({ $hasLabel, $hasHint, $valid }) => ($hasLabel ? 1 : 0) + ($hasHint ? 1 : 0) + (!$valid ? 1 : 0)}) {
        margin-bottom: var(--spacing-half);
    }
`;
