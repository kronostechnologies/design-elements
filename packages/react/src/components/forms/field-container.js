"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importDefault(require("react"));
var styled_components_1 = __importDefault(require("styled-components"));
var invalid_field_1 = require("../feedbacks/invalid-field");
var label_1 = require("./label");
var StyledDiv = styled_components_1.default.div(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  margin: 0 0 1.5rem;\n\n  input,\n  select,\n  textarea {\n    border-color: ", ";\n  }\n\n  &:focus {\n    border-color: ", ";\n  }\n"], ["\n  margin: 0 0 1.5rem;\n\n  input,\n  select,\n  textarea {\n    border-color: ", ";\n  }\n\n  &:focus {\n    border-color: ", ";\n  }\n"])), function (props) { return (props.valid ? 'rgb(217, 221, 226)' : 'rgb(164, 12, 46)'); }, function (props) { return (props.valid ? 'rgb(0, 128, 165)' : 'rgb(164, 12, 46)'); });
var FieldContainer = function (_a) {
    var children = _a.children, fieldId = _a.fieldId, label = _a.label, valid = _a.valid, validMsg = _a.validMsg, props = __rest(_a, ["children", "fieldId", "label", "valid", "validMsg"]);
    return (<StyledDiv {...props} valid={valid}>
        {label && (<label_1.Label forId={fieldId}>
                {label}
            </label_1.Label>)}

        {children}

        {!valid &&
        <invalid_field_1.InvalidField controlId={fieldId} feedbackMsg={validMsg}/>}
    </StyledDiv>);
};
exports.FieldContainer = FieldContainer;
var templateObject_1;
