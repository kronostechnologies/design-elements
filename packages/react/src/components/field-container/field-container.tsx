import { FunctionComponent, PropsWithChildren } from 'react';
import styled from 'styled-components';
import { Theme } from '../../themes/tokens/theme';
import { useDeviceContext } from '../device-context-provider/device-context-provider';
import { InvalidField } from '../feedbacks/invalid-field';
import { Label } from '../label/label';
import { TooltipProps } from '../tooltip/tooltip';

interface StyledDivProps {
    theme: Theme;
    hasLabel: boolean;
    hasHint: boolean;
    valid: boolean;
    noMargin?: boolean;
}

const StyledDiv = styled.div<StyledDivProps>`
    margin: ${({ noMargin }) => (noMargin ? '0' : '0 0 var(--spacing-3x)')};

    input,
    select,
    textarea {
        border-color: ${({ theme, valid }) => (valid ? theme.greys['dark-grey'] : theme.notifications['alert-2.1'])};
    }

    &:focus {
        border-color: ${({ theme, valid }) => (valid ? theme.main['primary-1.1'] : theme.notifications['alert-2.1'])};
    }

    > :nth-child(${({ hasLabel, hasHint, valid }) => (hasLabel ? 1 : 0) + (hasHint ? 1 : 0) + (!valid ? 1 : 0)}) {
        margin-bottom: var(--spacing-half);
    }
`;

const StyledHint = styled.span<{ isMobile: boolean }>`
    color: ${(props) => props.theme.greys['dark-grey']};
    display: block;
    font-size: ${({ isMobile }) => (isMobile ? '0.875rem' : '0.75rem')};
    font-weight: var(--font-normal);
    letter-spacing: 0.02rem;
    line-height: ${({ isMobile }) => (isMobile ? '1.5rem' : '1.25rem')};
`;

export interface FieldContainerProps {
    className?: string;
    fieldId: string;
    hint?: string;
    label?: string;
    noInvalidFieldIcon?: boolean;
    noMargin?: boolean;
    required?: boolean;
    tooltip?: TooltipProps;
    valid: boolean;
    validationErrorMessage: string;
}

export const FieldContainer: FunctionComponent<PropsWithChildren<FieldContainerProps>> = ({
    children,
    className,
    fieldId,
    hint,
    label,
    noInvalidFieldIcon,
    noMargin,
    required,
    tooltip,
    valid,
    validationErrorMessage,
    ...props
}) => {
    const { isMobile } = useDeviceContext();

    return (
        <StyledDiv
            className={className}
            noMargin={noMargin}
            hasLabel={!!label}
            hasHint={!!hint}
            valid={valid}
            {...props /* eslint-disable-line react/jsx-props-no-spreading */}
        >
            {label && <Label id={`${fieldId}_label`} forId={fieldId} tooltip={tooltip} required={required}>{label}</Label>}
            {hint && <StyledHint id={`${fieldId}_hint`} isMobile={isMobile}>{hint}</StyledHint>}
            {!valid && (
                <InvalidField
                    data-testid="text-input-error-msg"
                    controlId={fieldId}
                    feedbackMsg={validationErrorMessage}
                    noIcon={noInvalidFieldIcon}
                />
            )}
            {children}
        </StyledDiv>
    );
};
