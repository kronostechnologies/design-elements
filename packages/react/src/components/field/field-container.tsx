import { FunctionComponent, PropsWithChildren, useMemo } from 'react';
import styled from 'styled-components';
import { useId } from '../../hooks/use-id';
import { FieldControlContext } from './context';
import { InvalidFieldMessage } from '../feedbacks/invalid-field-message';
import { Label } from '../label/label';
import { Hint } from '../hint/hint';
import { FieldContainerProps } from './types';
import { getAriaDescribedby, getAriaLabel, getAriaLabelledby, getSlotIds } from './utils';

interface StyledDivProps {
    $hasLabel: boolean;
    $hasHint: boolean;
    $valid: boolean;
    $noMargin?: boolean;
}

const StyledDiv = styled.div<StyledDivProps>`
    margin: ${({ $noMargin }) => ($noMargin ? '0' : '0 0 var(--spacing-3x)')};

    input,
    select,
    textarea {
        border-color: ${({ theme, $valid }) => ($valid ? theme.component['field-input-border-color'] : theme.component['field-input-error-border-color'])};
    }

    > :nth-child(${({ $hasLabel, $hasHint, $valid }) => ($hasLabel ? 1 : 0) + ($hasHint ? 1 : 0) + (!$valid ? 1 : 0)}) {
        margin-bottom: var(--spacing-half);
    }
`;

export const FieldContainer: FunctionComponent<PropsWithChildren<FieldContainerProps>> = ({
    id: providedId,
    ariaLabel: providedAriaLabel,
    ariaLabelledby: providedAriaLabelledby,
    ariaDescribedby: providedAriaDescribedby,
    children,
    className,
    disabled = false,
    hint,
    label,
    noInvalidFieldIcon = false,
    noMargin,
    required = false,
    tooltip,
    valid = true,
    validationErrorMessage,
    ...props
}) => {
    const fieldId = useId(providedId);

    const slotIds = getSlotIds(fieldId, label, hint, (validationErrorMessage && !valid));
    const ariaLabel = getAriaLabel(label, providedAriaLabel, providedAriaLabelledby);
    const ariaLabelledby = getAriaLabelledby(slotIds, providedAriaLabelledby);
    const ariaDescribedby = getAriaDescribedby(slotIds, providedAriaDescribedby);

    const contextValues = useMemo(() => ({
        id: fieldId,
        ariaLabel,
        ariaLabelledby,
        ariaDescribedby,
        valid,
        hint,
        required,
        disabled,
    }), [fieldId, ariaLabel, ariaLabelledby, ariaDescribedby, valid, hint, required, disabled]);

    return (
        <FieldControlContext.Provider value={contextValues}>
            <StyledDiv
                className={className}
                $noMargin={noMargin}
                $hasLabel={!!label}
                $hasHint={!!hint}
                $valid={valid}
                {...props /* eslint-disable-line react/jsx-props-no-spreading */}
            >
                {label && (
                    <Label
                        id={slotIds.label}
                        htmlFor={fieldId}
                        tooltip={tooltip}
                        required={required}
                    >
                        {label}
                    </Label>
                )}
                {hint && (
                    <Hint id={slotIds.hint}>
                        {hint}
                    </Hint>
                )}
                {!valid && (
                    <InvalidFieldMessage id={slotIds.invalid} noInvalidFieldIcon={noInvalidFieldIcon}>
                        {validationErrorMessage}
                    </InvalidFieldMessage>
                )}
                {children}
            </StyledDiv>
        </FieldControlContext.Provider>
    );
};
