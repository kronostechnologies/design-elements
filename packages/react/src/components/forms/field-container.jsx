import React from 'react';
import styled from 'styled-components';

import Label from './label';
import InvalidField from '../feedbacks/invalid-field';

const FieldContainer = styled.div`
    margin: 0 0 1.5rem;

    input,
    select,
    textarea {
    border-color: ${props => props.valid ? 'rgb(217, 221, 226)' : 'rgb(164, 12, 46)'};

    &:focus{
        border-color: ${props => props.valid ? 'rgb(0, 128, 165)' : 'rgb(164, 12, 46)'};
    }
`;

export default ({ children, fieldId, label, valid, validMsg, ...props }) => (
    <FieldContainer {...props} valid={valid}>
        {label &&
             <Label forId={fieldId}>
                 {label}
             </Label>
        }

        {children}

        {!valid &&
         <InvalidField controlId={fieldId} feedbackMsg={validMsg} />
        }
    </FieldContainer>
);
