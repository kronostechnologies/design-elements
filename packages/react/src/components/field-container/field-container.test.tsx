import React from 'react';
import { renderWithProviders } from '../../test-utils/renderer';
import { FieldContainer } from './field-container';

describe('Field Container', () => {
    test('normal', () => {
        const tree = renderWithProviders(
            <FieldContainer valid {...defaultProps}>
                Children
            </FieldContainer>,
        );

        expect(tree).toMatchSnapshot();
    });

    test('invalid', () => {
        const tree = renderWithProviders(
            <FieldContainer valid={false} {...defaultProps}>
                Children
            </FieldContainer>,
        );

        expect(tree).toMatchSnapshot();
    });

    test('has no margin', () => {
        const tree = renderWithProviders(
            <FieldContainer valid disableMargin {...defaultProps}>
                Children
            </FieldContainer>,
        );

        expect(tree).toMatchSnapshot();
    });
});

const defaultProps = {
    fieldId: 'test-id',
    label: 'test label',
    validationErrorMessage: 'This text area input is invalid',
};
