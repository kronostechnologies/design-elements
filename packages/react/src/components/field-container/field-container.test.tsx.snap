// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field Container matches snapshot 1`] = `
.c2 {
  color: #1B1C1E;
  display: block;
  font-size: 0.75rem;
  font-weight: var(--font-normal);
  -webkit-letter-spacing: 0.02rem;
  -moz-letter-spacing: 0.02rem;
  -ms-letter-spacing: 0.02rem;
  letter-spacing: 0.02rem;
  line-height: 1.5rem;
  margin: 0;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
}

input + .c1 {
  margin-left: var(--spacing-half);
}

.c0 {
  margin: 0 0 var(--spacing-3x);
}

.c0 input,
.c0 select,
.c0 textarea {
  border-color: #60666E;
}

.c0 > :nth-child(1) {
  margin-bottom: var(--spacing-quarter);
}

<div
  class="c0"
>
  <label
    class="c1 c2"
    for="test-id"
    id="test-id_label"
  >
    test label
  </label>
  Children
</div>
`;

exports[`Field Container should have invalid styles given valid prop is set to false 1`] = `
.c3 {
  color: #CD2C23;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 0.75rem;
  font-weight: var(--font-normal);
  -webkit-letter-spacing: 0.02rem;
  -moz-letter-spacing: 0.02rem;
  -ms-letter-spacing: 0.02rem;
  letter-spacing: 0.02rem;
  line-height: 1.25rem;
}

.c4 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-right: var(--spacing-base);
}

.c2 {
  color: #1B1C1E;
  display: block;
  font-size: 0.75rem;
  font-weight: var(--font-normal);
  -webkit-letter-spacing: 0.02rem;
  -moz-letter-spacing: 0.02rem;
  -ms-letter-spacing: 0.02rem;
  letter-spacing: 0.02rem;
  line-height: 1.5rem;
  margin: 0;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
}

input + .c1 {
  margin-left: var(--spacing-half);
}

.c0 {
  margin: 0 0 var(--spacing-3x);
}

.c0 input,
.c0 select,
.c0 textarea {
  border-color: #CD2C23;
}

.c0 > :nth-child(2) {
  margin-bottom: var(--spacing-quarter);
}

<div
  class="c0"
>
  <label
    class="c1 c2"
    for="test-id"
    id="test-id_label"
  >
    test label
  </label>
  <span
    class="c3"
    data-testid="invalid-field"
    id="test-id_invalid"
  >
    <svg
      class="c4"
      color="currentColor"
      focusable="false"
      height="16"
      width="16"
    />
    This text area input is invalid
  </span>
  Children
</div>
`;

exports[`Field Container should not have margins given noMargin prop is set to true 1`] = `
.c2 {
  color: #1B1C1E;
  display: block;
  font-size: 0.75rem;
  font-weight: var(--font-normal);
  -webkit-letter-spacing: 0.02rem;
  -moz-letter-spacing: 0.02rem;
  -ms-letter-spacing: 0.02rem;
  letter-spacing: 0.02rem;
  line-height: 1.5rem;
  margin: 0;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
}

input + .c1 {
  margin-left: var(--spacing-half);
}

.c0 {
  margin: 0;
}

.c0 input,
.c0 select,
.c0 textarea {
  border-color: #60666E;
}

.c0 > :nth-child(1) {
  margin-bottom: var(--spacing-quarter);
}

<div
  class="c0"
>
  <label
    class="c1 c2"
    for="test-id"
    id="test-id_label"
  >
    test label
  </label>
  Children
</div>
`;
